src/linker/Linker.Steps/SweepStep.cs

```diff
@@ -728,16 +553,345 @@ protected virtual void ElementRemoved (IMetadataTokenProvider element)
 		{
 		}
 
-		protected virtual void ReferenceRemoved (AssemblyDefinition assembly, AssemblyNameReference reference)
+		protected virtual void InterfaceRemoved (TypeDefinition type, InterfaceImplementation iface)
 		{
 		}
 
-		protected virtual void InterfaceRemoved (TypeDefinition type, InterfaceImplementation iface)
+		protected virtual void CustomAttributeUsageRemoved (ICustomAttributeProvider provider, CustomAttribute attribute)
 		{
 		}
 
-		protected virtual void CustomAttributeUsageRemoved (ICustomAttributeProvider provider, CustomAttribute attribute)
+		struct AssemblyReferencesCorrector
 		{
+			readonly AssemblyDefinition assembly;
+			readonly DefaultMetadataImporter importer;
+
+			HashSet<TypeReference> updated;
+
+			public AssemblyReferencesCorrector (AssemblyDefinition assembly)
+			{
+				this.assembly = assembly;
+				this.importer = new DefaultMetadataImporter (assembly.MainModule);
+
+				updated = null;
+			}
+
+			public void Process ()
+			{
+				updated = new HashSet<TypeReference> ();
+
+				UpdateCustomAttributesTypesScopes (assembly);
+
+				foreach (var module in assembly.Modules)
+					UpdateCustomAttributesTypesScopes (module);
+
+				var mmodule = assembly.MainModule;
+				if (mmodule.HasTypes) {
+					foreach (var type in mmodule.Types) {
+						UpdateScopes (type);
+					}
+				}
+
+				if (mmodule.HasExportedTypes)
+					UpdateTypeScope (mmodule.ExportedTypes);
+
+				updated = null;
+			}
+
+			void UpdateScopes (TypeDefinition typeDefinition)
+			{
+				UpdateCustomAttributesTypesScopes (typeDefinition);
+
+				if (typeDefinition.BaseType != null)
+					UpdateTypeScope (typeDefinition.BaseType);
+
+				if (typeDefinition.HasInterfaces) {
+					foreach (var iface in typeDefinition.Interfaces) {
+						UpdateCustomAttributesTypesScopes (iface);
+						UpdateTypeScope (iface.InterfaceType);
+					}
+				}
+
+				if (typeDefinition.HasEvents) {
+					foreach (var e in typeDefinition.Events) {
+						UpdateCustomAttributesTypesScopes (e);
+						// e.EventType is not saved
+					}
+				}
+
```

Comment:
What about generic parameters on types? Specifically constrains and custom attributes?

Reply:
Yep, missed that one.